Bcm2836 irq

andreamerello opened the pull request:
Fixed local IRQ base irq number was incorrect.
RPI2 boots now :)

_NOTE_: The of_iomap of 2nd register address space works only appending the DTB to the zImage, selecting CONFIG_ARM_APPENDED_DTB in kernel CFG: strong suspect that the bootloader (at least mine) pretends to patch the DTB while loading it, messing it up.

IPI send implemented (hopefully) in complementary manner to the current IPI receive code. Yet to be tested.

lkundrak committed:
mailbox: Enable BCM2835 mailbox support

Implement BCM2835 mailbox support as a device registered with the
general purpose mailbox framework. Implementation based on commits by
Lubomir Rintel [1], Suman Anna and Jassi Brar [2] on which to base the
implementation.

[1] http://lists.infradead.org/pipermail/linux-rpi-kernel/2013-April/000528.html
[2] http://lists.infradead.org/pipermail/linux-rpi-kernel/2013-May/000546.html

v2: Squashed Craig's work for review, carried over to new version of
    Mailbox framework (changes by Lubomir)

v3: Fix multi-line comment style.  Refer to the documentation by
    filename.  Only declare one MODULE_AUTHOR.  Alphabetize includes.
    Drop some excessive dev_dbg()s (changes by anholt).

v4: Use the new bcm2835_peripheral_read_workaround(), drop the
    unnecessary wmb()s, make the messages be a pointer to u32, rather
    than u32-cast-as-pointer, fold in small static functions, drop
    extra error messages, clean up sizeof() arg for malloc, disable
    interrupts on unload (changes by anholt).

v5: Minor style fix.  Drop peripheral_read_workaround() since it's
    been noted that readl() covers the non-interrupt half of it, and I
    found that we would need to resolve read ordering in the interrupt
    handler anyway (changes by anholt).

Signed-off-by: Lubomir Rintel <lkundrak@v3.sk>
Signed-off-by: Craig McGeachie <slapdau@yahoo.com.au>
Signed-off-by: Suman Anna <s-anna@ti.com>
Signed-off-by: Jassi Brar <jassisinghbrar@gmail.com>
Signed-off-by: Eric Anholt <eric@anholt.net>
Cc: Stephen Warren <swarren@wwwdotorg.org>
Cc: Jassi Brar <jassisinghbrar@gmail.com>
Cc: Lee Jones <lee.jones@linaro.org>
Cc: linux-rpi-kernel@lists.infradead.org
Cc: linux-arm-kernel@lists.infradead.org

Commit:
dt-bindings: Add vendor prefix for Raspberry Pi

Since the prefix is already in use, we need to add it in the
vendor list.

Reviewed-by: Stephen Warren <swarren@wwwdotorg.org>
Acked-by: Rob Herring <robh@kernel.org>
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>
Signed-off-by: Lee Jones <lee.jones@linaro.org>

Commit:
dt-bindings: Add root properties for Raspberry Pi

This patch adds root compatible properties for the following boards:
- Raspberry Pi Model A
- Raspberry Pi Model A+
- Raspberry Pi Model B
- Raspberry Pi Model B (no P5)
- Raspberry Pi Model B rev2
- Raspberry Pi Model B+
- Raspberry Pi Compute Module

Reviewed-by: Stephen Warren <swarren@wwwdotorg.org>
Acked-by: Rob Herring <robh@kernel.org>
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>
Signed-off-by: Lee Jones <lee.jones@linaro.org>

Commit:
ARM: bcm2835: Add header file for pinctrl constants

This new header file defines pincontrol constants to use
from bcm2835 DTS files for pincontrol properties option.

Reviewed-by: Stephen Warren <swarren@wwwdotorg.org>
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>
Signed-off-by: Lee Jones <lee.jones@linaro.org>

Commit:
ARM: bcm2835: Use pinctrl header

This patch converts all bcm2835 dts and dtsi files to use the pinctrl
header file.

Reviewed-by: Stephen Warren <swarren@wwwdotorg.org>
Signed-off-by: Stefan Wahren <stefan.wahren@i2se.com>
Signed-off-by: Lee Jones <lee.jones@linaro.org>

anholt committed:
ARM: Add Kconfig support for bcm2836.

This port to the Raspberry Pi 2 comes up with SD card and network
working.  It doesn't yet support SMP, the ARM local timer, or PMU
events, and serial is fixed in a follow-on patch.

v2: Implement Arnd's feedback to not split to ARCH_BCM2836, and
    instead use more conditionals in ARCH_BCM2835.  Also reduce diff
    between 2835 and 2836.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
ARM: Make a copy of the 2835 dts for the 2836.

This is not hooked up to anything yet, but it means that our later
2836-specific changes will be more obvious.  The bcm2836-rpi-2-b.dts
actually comes from bcm2835-rpi-2-b-plus, which apparently is the
closest.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
ARM: Update the device trees for 2836.

The bus address for peripherals comes from the platform.h diff between
2708 and 2709, and the CPU's DT definition is based on the 2709 DT.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
ARM: Add MAINTAINERS for 2836.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
ARM: bcm2835: Add the mailbox to the device tree.

Signed-off-by: Eric Anholt <eric@anholt.net>
Acked-by: Lee Jones <lee.jones@linaro.org>

anholt committed:
ARM: bcm2835: Add the necessary firmware power domain information to the DT.

This gets the USB power domain turned on so that DWC2 can probe, even
if the bootloader didn't turn it on for us (which is the case if you
use the RPi firmware without chainloading U-Boot).

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
ARM: bcm2835: Add the mailbox power channel as a power domain provider.

This gives us control of the 3 power domains provided through the
firmware's mailbox.

v2: Convert from a function call to enable USB to an actual power
    domain so that USB's need for power can be expressed only in the
    devicetree, and add support for the other power domains at the
    same time. Move from drivers/mailbox/ to arch/arm/mach-*, where
    other power domain drivers appear to live.

v3: Fix some style, drop useless error message, rely on core return
    for probe defer, don't bother to devm_kfree() in probe error path.

v4: Fix copy-and-pasted bit twiddling for power off.  Fix return value
    of bcm_mbox_set_power() in case of repeated success.

Signed-off-by: Eric Anholt <eric@anholt.net>

Conflicts:
	arch/arm/mach-bcm/Kconfig
	arch/arm/mach-bcm/Makefile

anholt committed:
PM / Domains: If an OF node is found but no device probed yet, defer.

If we've declared a power domain in the OF, and the OF node is found
but the requested domain hasn't been registered on it yet, then we
probably have just tried to probe before the power domain driver has.
Defer our device's probe until it shows up.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
ARM: BCM2836: Add io map initialization for bcm2836.

baruchsiach committed:
ARM: bcm2835: Fix i2c0 node name

Device tree node names should contain the node's reg property address value.
The i2c0 node was apparently forgotten in commit 25b2f1bd0b7e0 (ARM: bcm2835:
node name unit address cleanup).

Acked-by: Stephen Warren <swarren@wwwdotorg.org>
Signed-off-by: Baruch Siach <baruch@tkos.co.il>
Signed-off-by: Lee Jones <lee.jones@linaro.org>

anholt committed:
ARM: BCM2835: Split peripheral definitions off to a common include.

The 2836 dts will inherit all of this, because it shares the same
peripherals as 2835.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
wip irq handler

anholt committed:
ARM: BCM2835: Register fixed clocks for uart in the DT.

We were previously relying on the fixed clock registration in
clk-bcm2835, but there doesn't seem to be any real reason to not just
define it in the DT (and for the 2836 port, I would have needed to
change the clock's physical address in clk-bcm2835.c).  Also, because
we weren't registering the apb_pclk in clk-bcm2835 as a clock device,
we were picking up the uart clock node as apb_pclk by accident.

This gets serial working on 2836.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
dt-bindings: Add root properties for Raspberry Pi 2.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
debug

anholt committed:
ARM: BCM2836: Try adding a bus mapping for the arm local control regs.

Still panics trying to of_iomap them.

andreamerello committed:
irq-bcm2835: fixed incorrect base for local IRQ regs

andreamerello committed:
irq-bcm2835: implement IPI send

andreamerello committed:
irq-bcm2835: enable IPIs for SMP

anholt commented:
Could you provide a Signed-off-by according to Documentation/SubmittingPatches?  In the comments here would be fine.

andreamerello commented:
I'm sorry, I just used to sign patches generated off my local git tree; I'm quite new to pull-request-based workflow and I didn't notice the signed-off-by was missing in commits themselves...

Please use this for the commits in this pull-request:
Signed-off-by: Andrea Merello <andrea.merello@gmail.com>

BTW I did some testing about IPIs code in this pull-request, and it seems working. I could see IPIs traveling across CPUs during boot by using debug prints (I'm booting all four cores with code in my bcm2386-smp branch, (but SMP boot does not reach completion yet)).

andreamerello committed:
irq-bcm2835: fix writing bitfield reg with int value

LOCAL_TIMER_INT_CONTROL0 (and well as it's siblings for other cores)
are actually bitfield.
We need to convert the integer irq index accordingly.

Signed-off-by: Andrea Merello <andrea.merello@gmail.com>

andreamerello committed:
irq-bcm2835: fix compare against irq_data ->irq instead of ->hwirq

Masking/unmasking IRQs has to be done in different manner depending
by the IRQ number (most notably depending if it falls inside the
per-core IRQ bank), and IRQ index off the first per-core IRQ must
eventually be extracted.

Static constants seems to refer to HW irq numbers, but they was
compared to irq_data->irq. Change in irq_data->hw_irq.

Signed-off-by: Andrea Merello <andrea.merello@gmail.com>

andreamerello committed:
irq-bcm2835: use already defined const instead of hardcoded value

No any functional change

Signed-off-by: Andrea Merello <andrea.merello@gmail.com>

anholt commented:
I think it's all merged to branches in my tree, now.
